src/xlogging.c

```diff
@@ -18,8 +38,123 @@ void xlogging_set_log_function(LOGGER_LOG log_function)
+    {
+        char* destination = result;
+        uint32_t destinationSize = TRACE_MAX_STACK_AS_STRING_SIZE;
+        
+        size_t destinationPos = 0;
+        destination[0] = '\0';
+
+        /*try fill destination up to destinationSize*/
+        char resultLine[TRACE_MAX_STACK_LINE_AS_STRING_SIZE];
+        void* stack[TRACE_MAX_STACK_FRAMES];
+
+        /*all following function calls are protected by the same SRW*/
+        AcquireSRWLockExclusive(&lockOverSymCalls);
+
+        uint16_t numberOfFrames = CaptureStackBackTrace(1, TRACE_MAX_STACK_FRAMES, stack, NULL);
+        HANDLE process = GetCurrentProcess();
+
+        SYMBOL_INFO_EXTENDED symbolExtended;
+        SYMBOL_INFO* symbol = &symbolExtended.symbol;
+
+        IMAGEHLP_LINE64 line;
+
+        symbol->MaxNameLen = TRACE_MAX_SYMBOL_SIZE;
+        symbol->SizeOfStruct = sizeof(SYMBOL_INFO);
+
+        for (uint32_t j = 0; j < numberOfFrames; j++)
+        {
+            DWORD64 address = (DWORD64)(stack[j]);
+            DWORD displacement = 0;
+
+            if (!SymFromAddr(process, address, NULL, symbol))
+            {
+                (void)strncat(destination, "SymFromAddr failed\n", destinationSize - destinationPos);
+            }
+            else
+            {
+                line.SizeOfStruct = sizeof(IMAGEHLP_LINE64);
+                if (SymGetLineFromAddr64(process, address, &displacement, &line))
+                {
+                    int snprintfResult = snprintf(resultLine, sizeof(resultLine), "!%s %s:%" PRIu32 "\n", symbol->Name, line.FileName, line.LineNumber);
+                    if (!(
+                        (snprintfResult >= 0) && /*the returned value is nonnegative [...]*/
+                        (snprintfResult < sizeof(resultLine)) /*[...] and less than n.*/
+                        ))
+                    {
+                        (void)strncat(destination, "snprintf failed\n", destinationSize - destinationPos);
+                    }
+                    else
+                    {
+                        (void)strncat(destination, resultLine, destinationSize - destinationPos);
+                    }
+                }
+                else
+                {
+                    int snprintfResult = snprintf(resultLine, sizeof(resultLine), "!%s Address 0x%" PRIX64 "\n", symbol->Name, line.Address);
+                    if (!(
+                        (snprintfResult >= 0) && /*the returned value is nonnegative [...]*/
+                        (snprintfResult < sizeof(resultLine)) /*[...] and less than n.*/
+                        ))
+                    {
+                        (void)strncat(destination, "snprintf failed\n", destinationSize - destinationPos);
+                    }
+                    else
+                    {
+                        (void)strncat(destination, resultLine, destinationSize - destinationPos);
+                    }
+                }
+            }
+            destinationPos = strlen(destination);
```

Comment:
>destinationPos = strlen(destination); [](start = 12, length = 37)

Consider: Calling strlen again is not really needed if one would add to destinationPos the number of chars that are strncat'd (the quantity is known in all cases - either COUNTOF("snprintf failed\n") or snprintfResult.

Reply:
I've considered and strlen is no longer called. This was a great suggestion!

---
In reply to: [276825923](https://github.com/Azure/azure-c-shared-utility/pull/302#discussion_r276825923) [](ancestors = 276825923)